celery_tasks_prioritizing/README.md

```diff
@@ -0,0 +1,86 @@
+# Prioritizing of the Celery tasks
+
+## Current state
+
+- 2 workers in prod, 1 worker in staging, no priority
+- in each worker we do:
+
+`celery --app="${APP}" worker --loglevel=${LOGLEVEL} --concurrency=1 --prefetch-multiplier=1`
+
+## What do we want to achieve
+
+- process the tasks depending on their priority
+- higher priority: tasks which should provide faster feedback - tasks triggered by user actions,
+  after user pushes/creates PR/comments, he should see that p-s processes this action as soon as possible (sets flags, ...)
+- lower priority: reporting Copr build start, TF results, actions triggered by other events that don't require
+  fast feedback to user
+
+| high-prio                      | low-prio                         |
+| ------------------------------ | -------------------------------- |
+| process_message                | run_copr_build_start_handler     |
+| run_copr_build_handler         | run_copr_build_end_handler       |
+| run_testing_farm_handler       | run_koji_build_report_handler    |
+| run_propose_downstream_handler | run_testing_farm_results_handler |
+| run_koji_build_handler         | run_installation_handler         |
+| run_pagure_pr_label_handler    | babysit_copr_build               |
+|                                | run_distgit_commit_handler       |
+
+## How to do it
+
+- from [FAQ](https://docs.celeryproject.org/en/master/faq.html#does-celery-support-task-priorities):
+  - Redis transport emulates priority support
+  - prioritize work by routing high priority tasks to different workers, this usually works better than per message priorities
+
+### Task priority
+
+- docs are not so clear:
+
+  - [priority](https://docs.celeryproject.org/en/latest/reference/celery.app.task.html?highlight=celery.app.task#celery.app.task.Task.priority)
+    attribute of the `Task` - default task priority
+  - [priority](https://docs.celeryproject.org/en/stable/reference/celery.app.task.html?highlight=priority#celery.app.task.Task.apply_async)
+    argument in `apply_async` method: The task priority, a number between 0 and 9 (defaults to the `priority` attribute of Task)
+
+  - [priority](https://docs.celeryproject.org/en/latest/userguide/calling.html#advanced-options) defined in Advanced
+    options of task calling - number between 0 and 255, where 255 is the highest priority, Redis (priority reversed, 0 is highest)
+
+#### [Redis message priorities](https://docs.celeryproject.org/en/latest/userguide/routing.html#redis-message-priorities)
+
+- there are 10 (0-9) priority levels
+- ```
+  app.conf.broker_transport_options = {
+    'queue_order_strategy': 'priority',
+    }
+  ```
+- creating `n` lists for each queue - by default 10 priority levels are consolidated to 4 levels
+  - e.g. queue named `celery` will be split into 4 queues:
+    ` ['celery0', 'celery3', 'celery6', 'celery9']`
+  - for more levels - `priority_steps` transport option
```

Comment:
How many levels do you suggest for us to use?

Reply:
I add the suggestion of plan to the doc, I would start with only defining a new queue and see how it will work and maybe later play also with these priority levels, as the documentation regarding it is confusing, found some issues (e.g. https://github.com/celery/celery/issues/4028 - mostly people recommend to emulate the priority on queue level with defining queues for different priorities)